File: examples-copy/docs/event-driven/events/event-system-ts/chat-service.ts
import { IEventBus } from "./event-system-0";
import { ChatStore } from "./stores-0";
import { Chat, Message } from "./types";
import { promises as fs } from "fs";
import * as path from "path";

class ChatService {
  private workspace: string = "tasks";

  constructor(private eventBus: IEventBus, private chatStore: ChatStore) {
    // Subscribe to commands
    this.eventBus.subscribe<StartNewChatCommand>(
      "StartNewChatCommand",
      this.handleCreateChat.bind(this)
    );

    this.eventBus.subscribe<UserSubmitMessageCommand>(
      "UserSubmitMessageCommand",
      this.handleUserMessage.bind(this)
    );

    this.eventBus.subscribe<SubmitInitialPromptCommand>(
      "SubmitInitialPromptCommand",
      this.handlePromptMessage.bind(this)
    );
  }

  private async handleStartNewChatCommand(
    command: StartNewChatCommand
  ): Promise<void> {
    const { taskId, subtaskId, metadata, correlationId } = command;
    const chatId = this.generateChatId();

    // Initialize chat object
    const chat: Chat = {
      id: chatId,
      taskId,
      subtaskId,
      messages: [],
      status: "active",
      createdAt: Date.now(),
      updatedAt: Date.now(),
      metadata,
    };
    this.chatStore.add(chat);

    // Create chat file
    await this.createChatFile(taskId, subtaskId, chatId);

    await this.eventBus.publish<ChatCreatedEvent>({
      type: "ChatCreated",
      taskId,
      subtaskId,
      chatId,
      timestamp: Date.now(),
      correlationId,
    });

    // Initialize first prompt based on subtask configuration
    const prompt = await this.generateInitialPrompt(taskId, subtaskId);
    const message: Message = {
      id: this.generateMessageId(),
      role: "@user", // This message is automatically sent by the system on behalf of user.
      content: prompt,
      timestamp: Date.now(),
      metadata: {
        taskId: chat.taskId,
        subtaskId: chat.subtaskId,
        isPrompt: true,
      },
    };

    // A trick to directly process the message. No need to wait.
    this.onMessageReceived(chat, message, correlationId);

    await this.eventBus.publish<ChatStartedEvent>({
      type: "ChatStarted",
      taskId,
      subtaskId,
      chatId,
      timestamp: Date.now(),
      correlationId,
    });
  }

  private async handleUserSubmitMessageCommand(
    command: UserSubmitMessageCommand
  ): Promise<void> {
    const { chatId, content, correlationId } = command;
    const chat = this.chats.get(chatId);

    if (!chat) {
      throw new Error(`Chat ${chatId} not found`);
    }

    const message: Message = {
      id: this.generateMessageId(),
      role: "@user",
      content,
      timestamp: Date.now(),
      metadata: {
        taskId: chat.taskId,
        subtaskId: chat.subtaskId,
      },
    };

    // 1. Save message to memory
    chat.messages.push(message);
    chat.updatedAt = Date.now();

    // 2. Emit MessageReceived event
    await this.eventBus.publish<MessageReceivedEvent>({
      type: "MessageReceived",
      chatId,
      message,
      timestamp: Date.now(),
      correlationId,
    });

    // 3. Save message to file
    const filePath = await this.saveMessageToFile(chat, message);

    // 4. Emit MessageSavedToFile event
    await this.eventBus.publish<MessageSavedToFileEvent>({
      type: "MessageSavedToFile",
      chatId,
      messageId: message.id,
      filePath,
      timestamp: Date.now(),
      correlationId,
    });

    // 5. Update chat status
    await this.eventBus.publish<ChatUpdatedEvent>({
      type: "ChatUpdated",
      chatId,
      lastMessageId: message.id,
      timestamp: Date.now(),
      correlationId,
    });

    // 6. Process message with agent
    await this.processMessageWithAgent(chat, message, correlationId);
  }

  private async onMessageReceived(
    chat: Chat,
    message: Message,
    correlationId: string
  ): Promise<void> {
    /**
     * Steps:
     * 1. Add to chat's meesage
     * 2. Update chat file
     * 3. If is agent message, do nothing and return
     * 3. If is user message
     *   - If has 'APPROVE' -> process the approval & generate the output -> output command
     *   - If not, generate the agent response.
     */

    chat.messages.push(message);

    // Save message to file
    const filePath = await this.saveMessageToChatFile(chat, message);

    // Should we emit here or emit at saveMessageToChatFile(...)
    await this.eventBus.publish<MessageSavedToChatFileEvent>({
      type: "MessageSavedToChatFileEvent",
      chatId,
      messageId: message.id,
      filePath,
      timestamp: Date.now(),
      correlationId,
    });

    // Process user message
    if (message.role === "@user") {
      // When user approve the work
      if (message.content.includes("APPROVE")) {
        await this.eventBus.publish<UserApprovedWorkEvent>({
          type: "UserApprovedWorkEvent",
          chatId: chat.id,
          timestamp: Date.now(),
          correlationId,
        });
        return;
      }

      // Continue to generate the agent response
      const response = await this.generateAgentResponse(
        chat,
        message,
        correlationId
      );
      await this.onMessageReceived(chat, response, correlationId);
    }
  }

  private async createChatFile() {
    // TODO
  }

  private async generateAgentResponse(
    chat: Chat,
    prompt: Message, // The last message of the chat
    correlationId: string
  ): Promise<Message> {
    // TODO: Generate agent response (this would be implemented by the agent service)
    const response: Message = {
      id: this.generateMessageId(),
      role: "@assistant",
      content: "Agent response placeholder",
      timestamp: Date.now(),
      metadata: {
        taskId: chat.taskId,
        subtaskId: chat.subtaskId,
      },
    };

    // Emit AgentResponseGenerated event
    await this.eventBus.publish<AgentResponseGeneratedEvent>({
      type: "AgentResponseGenerated",
      chatId: chat.id,
      response,
      timestamp: Date.now(),
      correlationId,
    });

    return response;
  }

  private getChatFilerPath(
    taskId: string,
    subtaskId: string,
    chatId: string
  ): string {
    return path.join(this.workspace, taskId, subtaskId, `chat_${chatId}.json`);
  }

  private async createChatFile(
    taskId: string,
    subtaskId: string,
    chatId: string
  ): Promise<void> {
    // await fs.mkdir(folderPath, { recursive: true });
    await fs.mkdir(this.getChatFilerPath(taskId, subtaskId, chatId), {
      recursive: true,
    });
  }

  private async saveMessageToChatFile(
    chat: Chat,
    message: Message
  ): Promise<string> {
    const { taskId, subtaskId, id: chatId } = chat;
    const filePath = this.getChatFilerPath(taskId, subtaskId, chatId);
    await fs.writeFile(filePath, JSON.stringify(message, null, 2));
    return filePath;
  }

  private async generateInitialPrompt(
    taskId: string,
    subtaskId: string
  ): Promise<string> {
    // TODO: This would be implemented based on the subtask configuration
    return "This is a sample initial prompt";
  }

  private generateChatId(): string {
    return `chat_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  private generateMessageId(): string {
    return `msg_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }
}

export default ChatService;

File: examples-copy/playground/events/chat_service.py
import asyncio
from datetime import datetime
from typing import Optional
from uuid import uuid4

from .event_bus import IEventBus
from .repositories import ChatRepository
from .type_defs import (
    Chat,
    ChatCreatedEvent,
    EventType,
    Message,
    MessageSavedToChatFileEvent,
    Role,
    StartNewChatCommand,
    UserApproveWorkEvent,
)
from .workspace_file_manager import WorkspaceFileManager


class ChatService:
    def __init__(
        self,
        event_bus: IEventBus,
        chat_repo: ChatRepository,
        workspace_manager: WorkspaceFileManager,
    ):
        self._workspace = 'tasks'
        self._event_bus = event_bus
        self._chat_repo = chat_repo
        self._workspace_manager = workspace_manager

        # Subscribe to commands
        self._event_bus.subscribe_async(
            EventType.START_NEW_CHAT_COMMAND, self._handle_start_new_chat_command
        )
        self._event_bus.subscribe_async(
            EventType.USER_SUBMIT_MESSAGE_COMMAND,
            self._handle_user_submit_message_command,
        )
        self._event_bus.subscribe_async(
            EventType.SUBMIT_INITIAL_PROMPT_COMMAND,
            self._handle_submit_initial_prompt_command,
        )

    async def _handle_start_new_chat_command(
        self, command: StartNewChatCommand
    ) -> None:
        chat_id = self._generate_chat_id()

        # Initialize chat object
        chat = Chat(
            id=chat_id,
            task_id=command.task_id,
            subtask_id=command.subtask_id,
            messages=[],
            status='active',
            created_at=datetime.now(),
            updated_at=datetime.now(),
            metadata=command.metadata,
        )
        self._chat_repo.save(chat)

        # Create chat file using workspace manager
        await self._workspace_manager.create_chat_file(
            command.task_id, command.subtask_id, chat_id
        )

        await self._event_bus.publish(
            ChatCreatedEvent(
                task_id=command.task_id,
                subtask_id=command.subtask_id,
                chat_id=chat_id,
                timestamp=datetime.now(),
                correlation_id=command.correlation_id,
            )
        )

        # Initialize first prompt based on subtask configuration
        prompt = await self._generate_initial_prompt(
            command.task_id, command.subtask_id
        )
        message = Message(
            id=self._generate_message_id(),
            role=Role.USER,
            content=prompt,
            timestamp=datetime.now(),
            metadata={
                'task_id': chat.task_id,
                'subtask_id': chat.subtask_id,
                'is_prompt': True,
            },
        )

        asyncio.create_task(
            self._on_message_received(chat, message, command.correlation_id)
        )

    async def _on_message_received(
        self, chat: Chat, message: Message, correlation_id: Optional[str]
    ) -> None:
        # Add message to chat
        chat.messages.append(message)

        # Save message to file using workspace manager
        file_path = await self._workspace_manager.save_message_to_chat_file(
            chat.id, message
        )

        # Emit event for message saved
        await self._event_bus.publish(
            MessageSavedToChatFileEvent(
                chat_id=chat.id,
                message_id=message.id,
                file_path=file_path,
                timestamp=datetime.now(),
                correlation_id=correlation_id,
            )
        )

        # Process user message if needed
        if message.role == Role.USER:
            if 'APPROVE' in message.content:
                await self._event_bus.publish(
                    UserApproveWorkEvent(
                        chat_id=chat.id,
                        timestamp=datetime.now(),
                        correlation_id=correlation_id,
                    )
                )
                return

            # Generate and process agent response
            response = await self._generate_agent_response(chat, message)
            await self._on_message_received(chat, response, correlation_id)

    def _generate_chat_id(self) -> str:
        return f'chat_{int(datetime.now().timestamp())}_{uuid4().hex[:8]}'

    def _generate_message_id(self) -> str:
        return f'msg_{int(datetime.now().timestamp())}_{uuid4().hex[:8]}'
